Add DescribeSensor integration tests

shane-axiom opened the pull request:
This pull request is a work in progress and should be discussed before being merged.

I was struggling to understand the expected/intended behavior of procedure description formats and DescribeSensor requests in SOS 4.4.x. I put together this integration test for [SOS-Test-Suite](https://github.com/52North/SOS-Test-Suite) to experiment and clearly manifest this behavior. I think it would be a good idea to add this test to the main SOS webapp tests in order to ensure that this behavior stays consistent in the future or detect when changes occur.

In summary, I found that:

* Procedures can only be inserted using formats which are identified as valid SOS 2.0 formats by `ProcedureEncoder`s. `SosHelper.checkFormat` checks the procedure description format against `CodingRepository.getInstance().getSupportedProcedureDescriptionFormats` using an SOS 2.0 ServiceOperatorKey. In general this means that only the (SOS 2.0 valid) URL formats can be used, and procedures can't be inserted using the (SOS 1.0 valid) mime type formats.
* Converters (e.g. `SensorML20SensorML101Converter`) generally allow different procedure description formats to be converted from one to another. However:
* SOS 2.0 `DescribeSensor` requests __cannot__ be made using mime type formats and SOS 1.0 `DescribeSensor` requests __cannot__ be made using URL formats because only `ProcedureEncoder.getSupportedProcedureDescriptionFormats` are checked (by service and version) for valid formats (via SosHelper.checkFormat and CodingRepository.getSupportedProcedureDescriptionFormats); only URL formats are provided for SOS 2.0 and only mime type formats are provided for SOS 1.0. __Converter formats are ignored__.
* Non-mime type formats immediately fail for SOS 1.0 `DescribeSensor` requests because `SosDescribeSensorOperatorV100.receive` always parses the format as a mime type and does not catch exceptions if the mime type parsing fails (e.g. URL formats cause an `IllegalArgumentException`.

Before this integration test is merged, can someone confirm that this is the intended behavior? I can understand limiting `DescribeSensor` requests by version to those formats identified as supported for the version, although it might be nice for the supported Converter formats to be checked as well. `SosDescribeSensorOperatorV100.receive`'s format handling (always assuming that the format is a mime type) seems like a bug to me.

CarstenHollmann commented on webapp/src/test/java/org/n52/sos/service/it/functional/DescribeSensorProcedureDescriptionFormatTest.java:
```diff
@@ -0,0 +1,385 @@
+
+    // Procedure inserted with SensorML 1.0.1 URL format can be requested with SOS 2.0
+    // using SensorML 2.0 URL format (http://www.opengis.net/sensorml/2.0)
+    public void testSos2DescribeSensorSensorML2Url() {
```
@Test missing?

CarstenHollmann commented:
@shane-axiom: Thanks for starting this integration test class.

The SOS 1.0.0 `DescribeSensor `allows only MimeType as `outputFormat `because the OGC SOS 1.0.0 XML schema defines the `outputFormat `as `MimeType`. 

I think the `IllegalArgumentException` is thrown in the [SosDecoderv100.parseDescribeSensor](https://github.com/52North/SOS/blob/develop/coding/sos-v100/src/main/java/org/n52/sos/decode/SosDecoderv100.java#L232) for XML requests and in the [DescribeSensorKvpDecoderv100.decode](https://github.com/52North/SOS/blob/develop/bindings/kvp/src/main/java/org/n52/sos/decode/kvp/v1/DescribeSensorKvpDecoderv100.java#L100) for KVP requests. But the decoder 
neither catch the `IllegalArgumentException`.

shane-axiom commented:
@CarstenHollmann Thanks for pointing out the missing `@Test` annotations and for your clarification and suggestions.

I have updated this pull request by:

* Adding KVP checks (in addition to POX checks) for all tests
* Fixing a problem with SOS 1.0 tests (previously was sending `swes:DescribeSensor` requests with version 1.0.0!)
* Only testing SOS 1.0.0 with non-mimetype format once to verify that it doesn't work
* Adding `MediaType.isMediaType(String)`, checking format in `SosDecoder100.parseDescribeSensor` and `DescribeSensorKvpDecoderv100.decode`, and throwing `InvalidOutputFormatParameterException` if a non-mimetype format is detected

SOS 1.0.0 POX DescribeSensor requests with non-mimetype outputFormats were already throwing an InvalidParameterException due to the XML schema validation, but KVP requests were throwing an `IllegalArgumentException` due to the non-mimetype parsing as `MediaType`. I added the same check (`MediaType.isMediaType`) to both for consistency.

I think this is ready for merge now if it looks good to you.
